File: language/src/env.rs
use crate::*;
use derivative::*;
use std::{
    cell::{RefCell, RefMut},
    collections::HashMap,
    rc::Rc,
};

#[derive(Debug, Clone, Default)]
pub struct Env(Rc<Inner>);

#[derive(Debug, Clone, Default)]
struct Inner {
    values: RefCell<HashMap<Id, (EnvKey, Any)>>,
    parent: Option<Env>,
}

impl Env {
    pub fn new() -> Self {
        Env::default()
    }

    pub fn with(prepare: impl FnOnce(&Env)) -> Env {
        let env = Env::new();
        prepare(&env);
        env
    }

    pub fn try_with<E>(
        prepare: impl FnOnce(&Env) -> std::result::Result<(), E>,
    ) -> std::result::Result<Env, E> {
        let env = Env::new();
        prepare(&env)?;
        Ok(env)
    }

    pub fn child(&self) -> Env {
        Env(Rc::new(Inner {
            parent: Some(self.clone()),
            ..Default::default()
        }))
    }

    pub fn parent(&self) -> Option<&Env> {
        self.0.parent.as_ref()
    }
}

thread_local! {
    static GLOBAL_ENV: Env = Env::new();
}

impl Env {
    pub fn global() -> Env {
        GLOBAL_ENV.with(Clone::clone)
    }

    pub fn clear_global() {
        Env::global().0.values.take();
    }

    pub fn is_global(&self) -> bool {
        std::ptr::eq(&self.0, &Env::global().0)
    }
}

impl Env {
    pub fn get<T: TypeInfo + Default>(&self, key: &EnvKey) -> T {
        self.values()
            .get(&key.id)
            .map(|(_, value)| value.cast::<T>().clone())
            .unwrap_or_default()
    }

    pub fn set<T: TypeInfo + Default>(&self, key: &EnvKey, value: T) {
        let key = key.clone();
        self.values().insert(key.id, (key, value.into()));
    }

    pub fn update<T: TypeInfo + Default, U>(
        &self,
        key: &EnvKey,
        update: impl FnOnce(&mut T) -> U,
    ) -> U {
        let mut value = self.get(key);
        let result = update(&mut value);
        self.set(key, value);
        result
    }
}

impl Env {
    fn values(&self) -> RefMut<HashMap<Id, (EnvKey, Any)>> {
        self.0.values.borrow_mut()
    }
}

#[derive(Debug, Clone, Derivative)]
#[derivative(PartialEq, Eq, Hash)]
pub struct EnvKey {
    id: Id,

    #[derivative(PartialEq = "ignore", Hash = "ignore")]
    pub visibility: EnvKeyVisibility,
}

impl EnvKey {
    pub fn new(visibility: EnvKeyVisibility) -> Self {
        EnvKey {
            id: Id::new(),
            visibility,
        }
    }
}

#[derive(Derivative, Clone)]
#[derivative(Debug)]
pub enum EnvKeyVisibility {
    /// The value cannot be `use`d by another environment.
    Private,

    /// The value can be `use`d by another environment. If the value already
    /// exists in the other environment, the provided `UseMergeFn` will be
    /// called to handle merging them. If the value doesn't exist, it will just
    /// be copied over.
    Public(UseMergeFn),
}

stored_closure!(pub struct UseMergeFn(&mut Any, &Any, &Env, &Stack) -> Result<()>);

impl UseMergeFn {
    pub fn merging<T: TypeInfo>(
        merge: impl Fn(&mut T, &T, &Env, &Stack) -> Result<()> + 'static,
    ) -> Self {
        UseMergeFn::new(move |current, new, env, stack| {
            merge(current.cast_mut(), new.cast(), env, stack)
        })
    }
}

impl Env {
    pub fn r#use(&self, other: &Env, stack: &Stack) -> Result<()> {
        let mut values = self.values();

        for (id, (key, new_value)) in other.values().iter() {
            match &key.visibility {
                EnvKeyVisibility::Public(merge) => match values.get_mut(id) {
                    Some((_, current_value)) => {
                        merge(current_value, new_value, self, stack)?;
                    }
                    None => {
                        values.insert(*id, (key.clone(), new_value.clone()));
                    }
                },
                EnvKeyVisibility::Private => {}
            }
        }

        Ok(())
    }
}

File: Language/Sources/Env.swift
public final class Env {
    private var items: [AnyEnvKey: Any]
    public var parent: Env?

    private init() {
        self.items = [:]
        self.parent = nil
    }

    public func child() -> Env {
        self.child { _ in }
    }

    public func child(prepare: (Env) throws -> Void) rethrows -> Env {
        let env = Env()
        env.parent = self
        try prepare(env)
        return env
    }
}

public extension Env {
    static let global = Env()

    var isGlobal: Bool {
        self === Env.global
    }
}

public extension Env {
    subscript<Key: EnvKey>(key: Key.Type) -> Key.Value? {
        get { self.items[AnyEnvKey(key)] as! Key.Value? }
        set { self.items[AnyEnvKey(key)] = newValue }
    }

    subscript(key: DynamicEnvKey) -> Any? {
        get { self.items[AnyEnvKey(key)] }
        set { self.items[AnyEnvKey(key)] = newValue }
    }

    func clear() {
        self.items.removeAll()
    }
}

public protocol EnvKey {
    associatedtype Value

    static var visibility: EnvKeyVisibility<Value> { get }
}

public struct DynamicEnvKey {
    public typealias Value = Any

    fileprivate let id: ID

    public let visibility: EnvKeyVisibility<Any>

    public init(visibility: EnvKeyVisibility<Any>) {
        self.id = ID()
        self.visibility = visibility
    }
}

private struct AnyEnvKey: Hashable {
    let id: ID
    let visibility: EnvKeyVisibility<Any>

    init<Key: EnvKey>(_ key: Key.Type) {
        self.id = ID(key)

        switch key.visibility {
        case .private:
            self.visibility = .private
        case .public(let merge):
            self.visibility = .public { erasedOld, erasedNew, env, stack in
                var old = erasedOld as! Key.Value
                let new = erasedNew as! Key.Value

                try merge(&old, new, env, stack)

                erasedOld = old
            }
        }
    }

    init(_ key: DynamicEnvKey) {
        self.id = key.id
        self.visibility = key.visibility
    }

    static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.id == rhs.id
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(self.id)
    }
}

public enum EnvKeyVisibility<Value> {
    /// The value cannot be `use`d by another environment.
    case `private`

    /// The value can be `use`d by another environment. If the value already
    /// exists in the other environment, the provided `UseMergeFn` will be
    /// called to handle merging them. If the value doesn't exist, it will just
    /// be copied over.
    case `public`(merge: Merge)

    public typealias Merge = (inout Value, Value, Env, Stack) throws -> Void
}

public extension Env {
    func use(_ other: Env, _ stack: Stack) throws {
        try self.use(other, self, stack)
    }

    func use(_ other: Env, _ env: Env, _ stack: Stack) throws {
        for (key, value) in other.items {
            switch key.visibility {
            case .public(let merge):
                if self.items.keys.contains(key) {
                    try merge(&self.items[key]!, value, env, stack)
                } else {
                    self.items[key] = value
                }
            case .private:
                break
            }
        }
    }
}
